pkgs/crypto/crypto.go

```diff
@@ -587,3 +595,95 @@ func EncryptPrivateKey(priv []byte, keyStorePassword string) ([]byte, error) {
 }
+
+func GetPubCommitsFromSharesData(reshare *wire.Reshare) ([]kyber.Point, error) {
+	suite := kyber_bls12381.NewBLS12381Suite()
+	signatureOffset := phase0.SignatureLength
+	pubKeysOffset := phase0.PublicKeyLength*len(reshare.OldOperators) + signatureOffset
+	sharesExpectedLength := EncryptedKeyLength*len(reshare.OldOperators) + pubKeysOffset
+	if len(reshare.Keyshares) != sharesExpectedLength {
+		return nil, fmt.Errorf("GetPubCommitsFromSharesData: shares data len is not correct, expected %d, actual %d", sharesExpectedLength, len(reshare.Keyshares))
+	}
+	pubKeys := utils.SplitBytes(reshare.Keyshares[signatureOffset:pubKeysOffset], phase0.PublicKeyLength)
+	// try to recover commits
+	var kyberPubShares []*share.PubShare
+	for i, pubk := range pubKeys {
+		blsPub := &bls.PublicKey{}
+		err := blsPub.Deserialize(pubk)
+		if err != nil {
+			return nil, err
+		}
+		v := suite.G1().Point()
+		err = v.UnmarshalBinary(blsPub.Serialize())
+		if err != nil {
+			return nil, err
+		}
+		kyberPubhare := &share.PubShare{
+			I: int(i),
+			V: v,
+		}
+		kyberPubShares = append(kyberPubShares, kyberPubhare)
+	}
+	pubPoly, err := share.RecoverPubPoly(suite.G1(), kyberPubShares, int(reshare.OldT), len(reshare.OldOperators))
+	if err != nil {
+		return nil, err
+	}
+	_, commits := pubPoly.Info()
+	return commits, nil
+}
+
+func GetSecretShareFromSharesData(reshare *wire.Reshare, opPrivateKey *rsa.PrivateKey, operatorID uint64) (*share.PriShare, error) {
+	suite := kyber_bls12381.NewBLS12381Suite()
+	signatureOffset := phase0.SignatureLength
+	pubKeysOffset := phase0.PublicKeyLength*len(reshare.OldOperators) + signatureOffset
+	sharesExpectedLength := EncryptedKeyLength*len(reshare.OldOperators) + pubKeysOffset
+	if len(reshare.Keyshares) != sharesExpectedLength {
+		return nil, fmt.Errorf("GetSecretShareFromSharesData: shares data len is not correct, expected %d, actual %d", sharesExpectedLength, len(reshare.Keyshares))
+	}
+	encryptedKeys := utils.SplitBytes(reshare.Keyshares[pubKeysOffset:], len(reshare.Keyshares[pubKeysOffset:])/len(reshare.OldOperators))
+	// try to decrypt private share
+	var kyberPrivShare *share.PriShare
+	for _, enck := range encryptedKeys {
+		prShare, err := rsaencryption.DecodeKey(opPrivateKey, enck)
+		if err != nil {
+			continue
+		}
+		secret := &bls.SecretKey{}
+		err = secret.SetHexString(string(prShare))
+		if err != nil {
+			return nil, err
+		}
+		// Check operator signature
+		var initVerify bool
+		initiatorPubKey, err := ParseRSAPubkey(reshare.InitiatorPublicKey)
+		if err != nil {
+			return nil, err
+		}
+		encInitPub, err := EncodePublicKey(initiatorPubKey)
+		if err != nil {
+			return nil, err
+		}
+		sigs := utils.SplitBytes(reshare.CeremonySigs, SignatureLength)
+		for _, sig := range sigs {
+			dataToVerify := make([]byte, len(secret.Serialize())+len(encInitPub))
+			copy(dataToVerify[:len(secret.Serialize())], secret.Serialize())
+			copy(dataToVerify[len(secret.Serialize()):], encInitPub)
```

Comment:
no need to serialize so many times
```suggestion
                       serialized := secret.Serialize()
			dataToVerify := make([]byte, len(serialized)+len(encInitPub))
			copy(dataToVerify[:len(serialized)], secret.Serialize())
			copy(dataToVerify[len(sserialized):], encInitPub)
```